metron-analytics/metron-profiler/src/main/java/org/apache/metron/profiler/bolt/ProfileBuilderBolt.java

```diff
@@ -0,0 +1,281 @@
+import org.apache.metron.common.bolt.ConfiguredProfilerBolt;
+import org.apache.metron.common.configuration.profiler.ProfileConfig;
+import org.apache.metron.common.dsl.ParseException;
+import org.apache.metron.profiler.ProfileMeasurement;
+import org.apache.metron.profiler.stellar.StellarExecutor;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * A bolt that is responsible for building a Profile.
+ *
+ * This bolt maintains the state required to build a Profile.  When the window
+ * period expires, the data is summarized as a ProfileMeasurement, all state is
+ * flushed, and the ProfileMeasurement is emitted.
+ *
+ * Each instance of this bolt is responsible for maintaining the state for a single
+ * Profile-Entity pair.
+ */
+public class ProfileBuilderBolt extends ConfiguredProfilerBolt {
+
+  protected static final Logger LOG = LoggerFactory.getLogger(ProfileBuilderBolt.class);
+
+  /**
+   * Executes Stellar code and maintains state across multiple invocations.
+   */
+  private StellarExecutor executor;
+
+  /**
+   * The number of seconds between when the Profile is flushed.
+   */
+  private int flushFrequency;
+
+  /**
+   * A ProfileMeasurement is created and emitted each window period.  A Profile
+   * itself is composed of many ProfileMeasurements.
+   */
+  private transient ProfileMeasurement measurement;
+
+  /**
+   * The definition of the Profile that the bolt is building.
+   */
+  private transient ProfileConfig profileConfig;
+
+  /**
+   * Parses JSON messages.
+   */
+  private transient JSONParser parser;
+
+  private OutputCollector collector;
+
+  /**
+   * @param zookeeperUrl The Zookeeper URL that contains the configuration data.
+   */
+  public ProfileBuilderBolt(String zookeeperUrl) {
+    super(zookeeperUrl);
+  }
+
+  /**
+   * Defines the frequency at which the bolt will receive tick tuples.  Tick tuples are
+   * used to control how often a profile is flushed.
+   */
+  @Override
+  public Map<String, Object> getComponentConfiguration() {
+    Config conf = new Config();
+    conf.put(Config.TOPOLOGY_TICK_TUPLE_FREQ_SECS, getFlushFrequency());
+    return conf;
+  }
+
+  @Override
+  public void prepare(Map stormConf, TopologyContext context, OutputCollector collector) {
+    super.prepare(stormConf, context, collector);
+    this.collector = collector;
+    this.parser = new JSONParser();
+  }
+
+  @Override
+  public void declareOutputFields(OutputFieldsDeclarer declarer) {
+    // once the time window expires, a complete ProfileMeasurement is emitted
+    declarer.declare(new Fields("measurement"));
+  }
+
+  @Override
+  public void execute(Tuple input) {
+    try {
+      doExecute(input);
+      collector.ack(input);
+
+    } catch (Throwable e) {
+      LOG.error("exception processing tuple: " + input, e);
+      collector.reportError(e);
```

Comment:
Yes, it should be acked whether there is an exception or not.  The main failure scenario is if someone creates a profile with a bad Stellar expression.  Retrying the tuple won't fix that.